kubeadm/flannel/Dockerfile

```diff
@@ -1,30 +1,39 @@
-ARG servercoreTag="ltsc2019"
-ARG cniVersion="0.8.5"
-ARG golangTag=windowsservercore-1809
+ARG BASE="mcr.microsoft.com/powershell:nanoserver-1809"
+ARG cniVersion="0.8.7"
+ARG flannelVersion="0.13.0"
 
-FROM golang:${golangTag} as builder
-ADD setup.go build/
-RUN go build -o build/setup.exe build/setup.go
-
-FROM mcr.microsoft.com/windows/servercore:${servercoreTag}
-SHELL ["powershell", "-NoLogo", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]
+FROM --platform=linux/amd64 golang as setup
+ENV GOOS=windows
+ENV GOARCH=amd64
+WORKDIR /build
+ADD setup.go .
+RUN go build -o setup.exe setup.go
 
+FROM --platform=linux/amd64 ubuntu as bins
+ARG flannelVersion
 ARG cniVersion
 
-RUN mkdir -force C:\k\flannel; \
-  pushd C:\k\flannel; \
-  curl.exe -LO https://github.com/coreos/flannel/releases/download/v0.12.0/flanneld.exe
+RUN apt-get update && apt-get install curl -y
+
+WORKDIR /flannel 
+RUN curl -Lo flanneld.exe https://github.com/coreos/flannel/releases/download/v${flannelVersion}/flanneld.exe
+
+WORKDIR /cni
+RUN curl -Lo cni.tgz https://github.com/containernetworking/plugins/releases/download/v${cniVersion}/cni-plugins-windows-amd64-v${cniVersion}.tgz
+RUN tar -xf cni.tgz
+RUN rm cni.tgz
+
+WORKDIR /utils
+RUN curl -Lo wins.exe https://github.com/rancher/wins/releases/download/v0.0.4/wins.exe
+RUN curl -Lo yq.exe https://github.com/mikefarah/yq/releases/download/2.4.1/yq_windows_amd64.exe
 
-ADD hns.psm1 /k/flannel
-COPY --from=builder /gopath/build/setup.exe /k/flannel/setup.exe
+FROM $BASE
 
-RUN mkdir C:\cni; \
-  pushd C:\cni; \
-  curl.exe -Lo cni.tgz https://github.com/containernetworking/plugins/releases/download/v${env:cniVersion}/cni-plugins-windows-amd64-v${env:cniVersion}.tgz; \
-  tar -xf cni.tgz; \
-  rm cni.tgz
+ADD hns.psm1 /k/flannel/hns.psm1
+COPY --from=setup /build/setup.exe /k/flannel/setup.exe
+COPY --from=bins /flannel/flanneld.exe /k/flannel/flanneld.exe
+COPY --from=bins /cni /cni
+COPY --from=bins /utils /utils
 
-RUN mkdir C:\utils; \
-  curl.exe -Lo C:\utils\wins.exe https://github.com/rancher/wins/releases/download/v0.0.4/wins.exe; \
-  curl.exe -Lo C:\utils\yq.exe https://github.com/mikefarah/yq/releases/download/2.4.1/yq_windows_amd64.exe; \
-  "[Environment]::SetEnvironmentVariable('PATH', $env:PATH + ';C:\utils', [EnvironmentVariableTarget]::Machine)"
+ENV PATH="C:\Program Files\PowerShell;C:\utils;C:\Windows\system32;C:\Windows;"
+USER ContainerAdministrator
```

Comment:
wins.exe doesn't work on ContainerUser. I don't know why, but it doesn't.

Reply:
I wonder if it is missing something in the container. 

@benmoss @marosset @claudiubelu do you have any ideas on why it might not work?

Reply:
`wins cli net get` returns empty string. so `$networkJson = wins cli net get | convertfrom-json` always fails.

Reply:
This happened to me when wins wasn't also installed on the host on which the container was running.

Reply:
Hmm, on second thought, that's not the same issue as I experienced. Wins can't even access the named pipe. Sorry.

Reply:
I am concerned on moving moving to nano server and switching to ContainerAdministrator as it is a big change for the way the image is used.  If we aren't able to track this down, we should go back to servercore so we can merge this and open a separate issue to track the change to nano without use of ContainerAdministrator.  Does that sound good?

Reply:
Ok. I will switch back to servercore.

Reply:
But I think that use ContainerAdministrator is not a big risk because we already use wins.exe and run code at host on SYSTEM account.

Reply:
@jsturtevant What do you think about build both: servercore (will work on User) and nanoserver (will work on Admin).
So we won't break current flow and will add possibility use nano (maybe with more risks)

Reply:
Rollback to servercore also affects build time.
flannel: from 3m (nanoserver) to 36m (servercore).
kube-proxy: from 25m (nanoserver) to 2h+ (servercore).

Reply:
I switched to nanoserver again because:
- we can't build kube-proxy for all k8s version for all windows version (3h is not enough for it)
- we run flannel.exe and kubeproxy.exe via wins as SYSTEM, so ContainerUser or ContainerAdministrator doesn't matter.

Reply:
@jsturtevant What do you think about using on nanoserver+ContainerAdmin in case when we can't automate build based on servercore? 
Do you think that using ContainerAdminstrator is unacceptable?

Reply:
I get your points though I still don't feel good about making that change.  @marosset do you have any thoughts or ideas on why wins.exe doesn't work in nanoserver?

Reply:
@marosset @jsturtevant a bit details: Why nanoserver doesn't support pipe mounting on ContainerUser.
wins.exe doesn't work because it has no access to \\.\pipe\rancher_wins when starts on ContainerUser.
It works only on ContainerAdministrator.

Reply:
@jsturtevant I found answers.
Default user for mcr.microsoft.com/windows/servercore is ContainerAdministrator
Default user for mcr.microsoft.com/windows/nanoserver is ContainerUser.

Proof:
![image](https://user-images.githubusercontent.com/12435135/99738634-04679100-2adc-11eb-83d2-d6d2b20fc5cc.png)

Reply:
And final message. I tried to build test image based on servercore
```
FROM mcr.microsoft.com/windows/servercore:1809
ADD https://github.com/rancher/wins/releases/download/v0.0.4/wins.exe /wins.exe
USER ContainerUser
```
Then I run it
```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: wins-test
  labels:
    tier: node
    app: wins-test
spec:
  selector:
    matchLabels:
      app: wins-test
  template:
    metadata:
      labels:
        tier: node
        app: wins-test
    spec:
      hostNetwork: true
      nodeSelector:
        kubernetes.io/os: 'windows'
      containers:
      - name: wins
        image: vleschenko/wins-test:user
        command:
        - ping 
        args:
        - -t
        - localhost
        volumeMounts:
        - name: wins
          mountPath: \\.\pipe\rancher_wins
        env:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
      volumes:
      - name: wins
        hostPath:
          path: \\.\pipe\rancher_wins
          type: null
```
When I connected to pod and tried run `wins cli net get` wins failed with the same error as on nanoserver - Access is denied.

Reply:
Thanks for digging into this!  Learned something new today :-).  I'll give it a final review today.

Reply:
Very interesting.